modules/navier_stokes/src/executioners/SIMPLE.C

```diff
@@ -0,0 +1,1106 @@
+    Real diag = mx(dof_id, dof_id);
+    rhs.add(dof_id, desired_value * diag);
+    mx.set(dof_id, dof_id, 2 * diag);
+  }
+}
+
+dof_id_type
+SIMPLE::findDoFID(const VariableName & var_name, const Point & point)
+{
+  // Find the element containing the point
+  auto point_locator = PointLocatorBase::build(TREE_LOCAL_ELEMENTS, _fe_problem.mesh());
+  point_locator->enable_out_of_mesh_mode();
+
+  const auto & variable = _fe_problem.getVariable(0, var_name);
+  unsigned int var_num = variable.sys().system().variable_number(var_name);
+
+  // We only check in the restricted blocks, if needed
+  const bool block_restricted =
+      variable.blockIDs().find(Moose::ANY_BLOCK_ID) == variable.blockIDs().end();
+  const Elem * elem =
+      block_restricted ? (*point_locator)(point, &variable.blockIDs()) : (*point_locator)(point);
+
+  // We communicate the results and if there is conflict between processes,
+  // the minimum cell ID is chosen
+  const dof_id_type elem_id = elem ? elem->id() : DofObject::invalid_id;
+  dof_id_type min_elem_id = elem_id;
+  comm().min(min_elem_id);
+
+  if (min_elem_id == DofObject::invalid_id)
+    mooseError("Variable ",
+               var_name,
+               " is not defined at ",
+               Moose::stringify(point),
+               "! Try alleviating block restrictions or using another point!");
+
+  return min_elem_id == elem_id ? elem->dof_number(variable.sys().number(), var_num, 0)
+                                : DofObject::invalid_id;
+}
+
+void
+SIMPLE::relaxSolutionUpdate(NonlinearSystemBase & system_in, Real relaxation_factor)
+{
+  // We will need the latest and the second latest solution for the relaxation
+  NumericVector<Number> & solution = *(system_in.system().current_local_solution.get());
+  NumericVector<Number> & solution_old = *(system_in.solutionPreviousNewton());
+
+  // The relaxation is just u = lambda * u* + (1-lambda) u_old
+  solution.scale(relaxation_factor);
+  solution.add(1 - relaxation_factor, solution_old);
+  solution.close();
+
+  if (_print_fields)
+  {
+    _console << "Pressure solution" << std::endl;
+    solution.print();
+    _console << "Pressure solution old" << std::endl;
+    solution_old.print();
+  }
+
+  // We will overwrite the old solution here
+  solution_old = solution;
+  system_in.setSolution(solution);
+  system_in.residualSetup();
+}
+
+std::vector<Real>
+SIMPLE::solveMomentumPredictor()
+{
+  // Temporary storage for the (flux-normalized) residuals form
+  // different momentum components
+  std::vector<Real> normalized_residuals;
+
+  // We can create this here with the assumption that every momentum component has the same number
+  // of dofs
+  auto zero_solution = _momentum_systems[0]->system().current_local_solution->zero_clone();
+
+  // If we use a segregated approach between momentum components as well, we need to solve
+  // the other equations. Luckily, the system matrix is exactly the same so we only need
+  // to compute right hand sides.
```

Comment:
is this comment still true?

Reply:
looks like you recompute the matrix every time below, as discussed prior

Reply:
Right, nice catch! Thanks!